File: site/wsgi/save.py
#!/usr/bin/env python

import sys, os, cgi, cgitb, re, json, base64, time, shutil, string
progpath = os.path.dirname(__file__)
sys.path.append(progpath)
import config

rootcachename = os.path.join(config.cachedir, 'rootcache')

def uri_without_query(uri):
  return uri.split('?', 1)[0]

def splituri(uri):
  return [p for p in uri_without_query(uri).split('/') if p]

def filename_from_uri(uri):
  return '/'.join(splituri(uri)[1:])

# Form fields:
# file="parent/dir/filename.ext" - The destination location.
# data="hello world.\nthis is a file.\n' - Data to write.
# key="888" - Short (insecure) password key to authorize user.
# source="other/dir/filename.ext" - File or directory to copy or move.
# sourcekey - For mv, authorization for the source file.
# mode=
#  "a" to append data,
#  "mv" to move source,
#  "rmtree" to remove dir.
#  "setkey" to set key.

def application(env, start_response):
  form = cgi.FieldStorage(
      fp=env['wsgi.input'], environ=env, keep_blank_values=True)
  request_uri = env['REQUEST_URI']
  host_name = env['HTTP_HOST']
  filename = form.getfirst("file", filename_from_uri(request_uri))
  data = form.getfirst("data", None)
  callback = form.getfirst("callback", None)
  sourcefile = form.getfirst("source", None)
  mode = form.getfirst("mode", None)
  conditional = form.getfirst("conditional", None)
  key = form.getfirst("key", '')
  sourcekey = form.getfirst("sourcekey", key)
  user = '.'.join(host_name.split('.')[:-2])
  domain = '.'.join(host_name.split('.')[-2:])
  origfilename = filename
  if len(user) is 0:
    user = None
  
  try:
    os.remove(rootcachename)
  except:
    pass
  
  def jsonout(dict):
    if callback is not None:
      return callback + '(' + json.dumps(dict) + ')'
    else:
      return json.dumps(dict)
 
  class ImmediateReturnException(Exception):
    def __init__(self, message, json):
      super(ImmediateReturnException, self).__init__(message)
      self.json = json

  def errorexit(s):
    raise ImmediateReturnException(s, exitdata({'error': s}))
  
  def exitdata(d):
    start_response('200 OK', [
      ('Content-Type', 'text/javascript')
    ])
    return jsonout(d)
  
  def filenameuser(fn):
    m = re.match(r'^([\w][\w\.\-]*)(?:/.*)?$', filename)
    if m is None:
      return None
    return m.group(1)
  
  def validkey(user, key):
    keydir = os.path.join(config.datadir, user, '.key')
    if not os.path.isdir(keydir):
      return True
    keys = os.listdir(keydir)
    if len(keys) == 0:
      return True
    for authorized in keys:
      if key.startswith(authorized[1:]):
        return True
  
  def setkey(user, oldkey, newkey):
    if oldkey == newkey:
      return
    keydir = os.path.join(config.datadir, user, '.key')
    try:
      if not os.path.isdir(keydir):
        checkreserveduser(user)
        os.makedirs(keydir)
      if oldkey is not None:
        keys = os.listdir(keydir)
        for authorized in keys:
          if oldkey.startswith(authorized[1:]):
            os.remove(os.path.join(keydir, authorized))
      if newkey is not None:
        keyfile = os.path.join(keydir, 'k' + newkey)
        open(keyfile, 'w').close()
    except OSError, exc:
      errorexit('Could not set key.')
  
  def checkreserveduser(user):
    if os.path.isdir(os.path.join(config.datadir, user)):
      return
    if user.lower() != user:
      errorexit('Username should be lowercase.')
    normalized = user.lower()
    if os.path.isdir(os.path.join(config.datadir, normalized)):
      errorexit('Username is reserved.')
    normalized = user.lower()
    if normalized != user and os.path.isdir(
          os.path.join(config.datadir, normalized)):
      errorexit('Username is reserved.')
    normalizedi = normalized.translate(string.maketrans('013456789', 'oieasbtbg'))
    if normalized != normalizedi and os.path.isdir(
          os.path.join(config.datadir, normalizedi)):
      errorexit('Username is reserved.')
    normalizedl = normalized.translate(string.maketrans('013456789', 'oleasbtbg'))
    if normalizedl != normalized and os.path.isdir(
          os.path.join(config.datadir, normalizedl)):
      errorexit('Username is reserved.')
    with open(os.path.join(progpath, 'bad-words.txt')) as f:
      badwords = f.read().splitlines()
    if any(word in badwords for word in [normalized, normalizedi, normalizedl]):
      errorexit('Username is reserved.')
    with open(os.path.join(progpath, 'bad-substrings.txt')) as f:
      badsubstrings = f.read().splitlines()
    if any(substring in word
         for word in [normalized, normalizedi, normalizedl]
         for substring in badsubstrings):
      errorexit('Username is reserved.')
    return

  try:
    
    # Validate params
    if data is not None and sourcefile is not None:
      errorexit('Cannot supply both data and source.')
    if mode is not None and mode not in (
        sourcefile and ['mv'] or
        data and ['a', 'setkey'] or
        (not data and not sourcefile) and ['rmtree', 'setkey']):
      errorexit('Illegal mode %s.' % mode + repr(data))
    if conditional is not None:
      if not re.match(r'^\d+(?:\.\d*)?$', conditional):
        errorexit('Illegal conditional %s.' % conditional)
      conditional = float(conditional)
    
    # Validate a good username
    if user is not None:
      if not re.match(r'^[A-Za-z]\w{2,}$', user):
        errorexit('Bad username "%s".' % user)
      filename = os.path.join(user, filename)
    # Validate a good filename: must have a leading dir.
    topdir = False
    if not re.match(r'^(?:[\w][\w\.\-]*)(?:/[\w][\w\.\-]*)+/?$', filename):
      # The rmtree and mv case are allowed to be a bare leading dir, subject
      # to additional checks.
      if (mode == 'setkey' or (data is None and (
          mode in ['rmtree', 'mv'] or (sourcefile is not None))) and
          re.match(r'^[\w][\w\\-]*/?$', filename)):
        topdir = True
      else:
        errorexit('Bad filename "%s".' % filename)
    absfile = os.path.join(config.datadir, filename)
    if not absfile.startswith(config.datadir):
      errorexit('Illegal filename "%s".' % filename)
    userdir = None
    if user:
      userdir = os.path.join(config.datadir, user)
    
    # Validate that the user's key matches the supplied key
    if not validkey(user, key):
      if not key:
        return exitdata({'error': 'Password protected.', 'needauth': 'key'})
      else:
        return exitdata({'error': 'Incorrect password.', 'needauth': 'key'})
    
    # Handle the setkey case
    if mode == 'setkey':
      if not topdir:
        errorexit('Can only set key on a top-level user directory.')
      setkey(user, key, data)
      return exitdata(data is None and {'keycleared': user} or {'keyset': user});
    
    # Handle the copy/move case
    if sourcefile is not None:
      if not re.match(r'^(?:[\w][\w\.\-]*)(?:/[\w][\w\.\-]*)*/?$', sourcefile):
        errorexit('Bad source filename "%s".' % sourcefile)
      sourceuser = filenameuser(sourcefile)
      abssourcefile = os.path.join(config.datadir, sourcefile)
      if not abssourcefile.startswith(config.datadir):
        errorexit('Illegal source filename "%s".' % sourcefile)
      if not os.path.exists(abssourcefile):
        errorexit('Source file "%s" does not exist.' % sourcefile)
      # Validate that only directories can be copied or moved to the top.
      if topdir and not os.path.isdir(abssourcefile):
        errorexit('Bad filename "%s".' % filename)
      # mv requires authorization on source dir
      if mode == 'mv':
        if not validkey(sourceuser, sourcekey):
          if not key:
            return exitdata({'error': 'Source password protected.', 'auth': 'key'})
          else:
            return exitdata({'error': 'Incorrect source password.', 'auth': 'key'})
      # create target parent directory
      if not os.path.isdir(os.path.dirname(absfile)):
        checkreserveduser(user)
        try:
          os.makedirs(os.path.dirname(absfile))
        except OSError, exc:
          errorexit('Could not create dir "%s".' % os.path.dirname(filename))
      # move case
      if mode == 'mv':
        if os.path.exists(absfile):
          errorexit('Cannot replace existing file "%s".' % filename)
        try:
          shutil.move(abssourcefile, absfile)
          try:
            os.removedirs(os.path.dirname(abssourcefile))
          except OSError, exc:
            pass
          # remove .key when moving top dir into deeper dir
          if topdir and filename != user:
            shutil.rmtree(os.path.join(absfile, '.key'))
        except OSError, exc:
          errorexit('Could not move "%s" to "%s".' % (sourcefile, filename))
      # copy case
      else:
        try:
          if os.path.isdir(abssourcefile):
            if os.path.exists(absfile):
              errorexit('Cannot overwrite existing directory "%s".' % filename)
            shutil.copytree(abssourcefile, absfile, ignore_patterns('.key'))
          else:
            shutil.copy2(abssourcefile, absfile)
        except OSError, exc:
          errorexit('Could not copy "%s" to "%s".' % (sourcefile, filename))
      try:
        os.utime(userdir, None)
      except OSError, exc:
        pass
      return exitdata({'saved' : filename})
    
    # Enforce the conditional request
    if conditional:
      if os.path.exists(absfile):
        mtime = os.stat(absfile).st_mtime
        if mtime > conditional:
          return exitdata({'error': 'Did not overwrite newer file.', 'newer': mtime})
    
    # Handle the delete case
    if not data:
      if not form.has_key('data'):
        errorexit('Missing data= cgi argument.')
      if os.path.exists(absfile):
        try:
          if os.path.isdir(absfile):
            if mode == 'rmtree':
              shutil.rmtree(absfile)
            else:
              os.rmdir(absfile)
          else:
            os.remove(absfile)
        except OSError, exc:
          errorexit('Could not remove "' + absfile + '".')
        try:
          os.removedirs(os.path.dirname(absfile))
        except OSError, exc:
          pass
      try:
        if userdir != absfile:
          os.utime(userdir, None)
      except OSError, exc:
        pass
      return exitdata({'deleted' : filename})
    
    # Validate data
    if len(data) >= 1024 * 1024:
      errorexit('Data too large.')
    
    # Handle the create/replace case
    if not os.path.isdir(os.path.dirname(absfile)):
      checkreserveduser(user)
      try:
        os.makedirs(os.path.dirname(absfile))
      except OSError, exc:
        errorexit('Could not create dir "' + os.path.dirname(filename) + '".')
    
    try:
      f = file(absfile, (mode == 'a' and 'a' or 'w'))
      f.write(data)
      f.flush()
      os.fsync(f.fileno());
      fs = os.fstat(f.fileno())
      f.close()
    except IOError, err:
      errorexit('Error writing file "' + absfile + '".')
    try:
      os.utime(userdir, None)
    except OSError, exc:
      pass
    
    return exitdata(
        {'saved' : filename, 'mtime': fs.st_mtime, 'size': fs.st_size})
  except ImmediateReturnException, e:
    return e.json

File: dev/save.js
var path = require('path');
var fs = require('fs');
var fsExtra = require('fs.extra');

exports.handleSave = function(req, res, app) {
  var data = req.param('data', null);
  var sourcefile = req.param('source', null);
  var mode = req.param('mode', null);
  var conditional = req.param('conditional', null);
  var key = req.param('key', null);
  var sourcekey = req.param('sourcekey', null);

  try {
    var domainIndex = req.host.indexOf(app.locals.config.host);
    if (domainIndex < 0) {
      errorExit('Domain not found in host: ' + app.locals.config.host);
    }

    // Adjust domainIndex for '.' separator
    domainIndex = (domainIndex > 0) ? domainIndex - 1 : domainIndex;

    var user = req.host.substring(0, domainIndex);
    var filename = req.param("file", filenameFromUri(req));
    var origfilename = filename;

    if (user == '') {
      user = null;
    }

    /*
    console.log({
      'user': user, 
      'key': key,
      'filename': filename,
      'data': data,
      'sourcefile': sourcefile,
      'conditional': conditional});
    */

    //
    // Validate parameters
    //

    if (data && sourcefile) {
      errorExit('Cannot supply both data and source');
    }

    if (mode) {
      switch (mode) {
        case 'mv':
          if (!sourcefile) {
	    errorExit('No source specified for mv');
          }
          break;
        case 'a':
          if (!data) {
	    errorExit('No data specified for append');
	  }
          break;
        case 'setkey': 
          if (!data && sourcefile) {
	    errorExit('Invalid parameters specified for setkey');
	  }
          break;
        case 'rmtree':
          if (data || sourcefile) {
	    errorExit('Either data or source specified for rmtree');
	  }
          break;
        default:
          errorExit('Unknown mode type');
      }      
    }

    if (conditional) {
      conditional = Date.parse(conditional);
      if (isNaN(conditional)) {
        errorExit('Illegal conditional');
      }
    }

    // validate username
    if (user) {
      if (!/^[A-Za-z]\w{2,}$/.test(user)) {
        errorExit('Bad username ' + user);
      }
      filename = path.join(user, filename);
    }

    var topdir = false;
    if (!/^(?:[\w][\w\.\-]*)(?:\/[\w][\w\.\-]*)+\/?$/.test(filename)) {
      if (mode == 'setkey' || 
	  (!data && (mode == 'rmtree' || mode == 'mv' || !sourcefile)) &&
	  /^[\w][\w\\-]*\/?$/.test(filename)) {
        topdir = true;
      }
      else {
        errorExit('Bad filename ' + filename);
      }
    }

    var absfile = makeAbsolute(filename, app);
    var userdir = null;
    if (user) {
      userdir = getUserHomeDir(user, app);
    }

    //
    // Validate that users key matches the supplied key
    //

    if (!isValidKey(user, key, app)) {
      var msg = (key) ? 'Incorrect password.' : 'Password protected.';
      res.json({'error': msg, 'needauth': 'key'});
      return;
    }

    //
    // Handle setkey
    //

    if (mode == 'setkey') {
      if (!topdir) {
        errorExit('Can only set key on a top-level user directory.');
      }

      doSetKey(user, key, data, res, app);
      res.json((data) ? {'keyset': user} : {'keycleared': user});
      return;
    }

    //
    // Handle the copy/move case
    //

    if (sourcefile) {
      if (!/^(?:[\w][\w\.\-]*)(?:\/[\w][\w\.\-]*)*\/?$/.test(sourcefile)) {
        errorExit('Bad source filename: ' + sourcefile);
      }

      sourceuser = filenameuser(sourcefile);
      var absSourceFile = makeAbsolute(sourcefile, app);
      if (!fs.existsSync(absSourceFile)) {
	errorExit('Source file does not exist. ' + sourcefile);
      }

      // Only directories can be copied or moved to the top
      if (topdir && !fs.statSync(absSourceFile).isDirectory()) {
	errorExit('Bad filename. ' + filename);
      }

      // mv requires authz on the source dir
      if (mode == 'mv') {
        if (!isValidKey(sourceuser, sourcekey, app)) {
	  var msg = (!key) ? 
	      'Source password protected.' : 'Incorrect source password.';
	  res.json({'error': msg, 'auth': 'key'});
	  return;
	}
      }

      // Create target parent directory if needed
      if (!fs.statSync(path.dirname(absfile)).isDirectory()) {
        checkReservedUser(user, app);
	try {
          fs.mkdirSync(path.dirname(absfile));
	}
	catch (e) {
          errorExit('Could not create dir: ' + path.dirname(filename));
	}
      }

      // move case
      if (mode == 'mv') {
	if (fs.existsSync(absfile)) {
          errorExit('Cannot replace existing file: ' + filename);
	}

	try {
	  fs.renameSync(absSourceFile, absfile);

	  // Cleanup directories if necessary
	  var dir = path.dirname(absSourceFile);
	  for (; dir ; dir = path.dirname(dir)) {
            try {
	      fs.rmdirSync(dir);
	    }
	    catch (e) {
	      // Failed to remove dir, assume not empty
	      break;
	    }
	  }

	  // Remove .key when moving top dir into deeper dir because we don't
	  // want to propagate password data
	  if (topdir && filename != user) {
            fsExtra.removeSync(path.join(absfile, '.key'));
	  }
	}
	catch (e) {
          errorExit('Could not move ' + sourcefile + ' to ' + filename);
	}
      }
      else {
	// Copy case
        try {
	  // Are we copying a directory?
          if (fs.stat(absSourceFile).isDirectory()) {
            if (fs.existsSync(absfile)) {
              errorExit('Cannot overwwrite existing directory ' + filename);
	    }

	    fsExtra.copySync(absSourceFile, absfile); 
	    // TODO: Need to ignore .key subdirs and contents
          }
          else {
            fsExtra.copySync(absSourceFile, absfile);
          }
        }
        catch (e) {
          errorExit('Could not copy ' + sourcefile + ' to ' + filename);
        }
      }

      touchUserDir(userdir);
      res.json({'saved': filename});
      return;
    }

    //
    // Enforce the conditional request if present
    //

    if (conditional) {
      if (fs.existsSync(absfile)) {
	mtime = Date.parse(fs.statSync(absfile).mtime);
        if (mtime > conditional) {
            res.json({'error': 'Did not overwrite newer file.', 
                      'newer': mtime});
            return;
        }
      }
    }

    // 
    // Handle the delete case
    //

    if (!data) {
        //if (!req.body.hasOwnProperty('data')) {
        //errorExit('Missing data= form field argument.');
        //}

      if (fs.existsSync(absfile)) {
        try {
          fsExtra.removeSync(absfile);
	}
	catch (e) {
          errorExit('Could not remove: ' + absfile);
	}

        try {
          removeDirsSync(path.dirname(absfile));
	}
	catch (e) { }
      }
	
      if (userdir != absfile) {
        touchUserDir(userdir);
      }

      res.json({'deleted' : filename});
      return;
    }

    // Validate data
    if (data.length > 1024 * 1024) {
      errorExit('Data too large.');
    }

    //
    // Finally handle the create/replace case
    //

    if (!fs.existsSync(path.dirname(absfile)) ||
	!fs.statSync(path.dirname(absfile)).isDirectory()) {
      checkReservedUser(user, app);
      try {
	fsExtra.mkdirsSync(path.dirname(absfile));
      }
      catch (e) {
        errorExit('Could not create dir: ' + path.dirname(filename));
      }
    }

    var statObj;
    try {
      var openMode = (mode == 'a') ? 'a' : 'w';
      fd = fs.openSync(absfile, openMode);

      writeStream = fs.createWriteStream(absfile, {flags: openMode});
      writeStream.write(data);
      writeStream.end();

      writeStream.on('close', function() {
        fs.fsyncSync(fd);
	fs.closeSync(fd);

	statObj = fs.statSync(absfile);
	touchUserDir(userdir);

	res.json({saved: filename, mtime: statObj.mtime, size: statObj.size});
      });
      return;
    }
    catch (e) {
      errorExit('Error writing file: ' + absfile);
    }

    return;
  }
  catch (e) {
    if (e instanceof ImmediateReturnError) {
      res.json(e.jsonObj);
    }
    else {
      throw e;
    }
  }
}

function errorExit(msg) {
  console.log('ERROREXIT: ' + msg);
  throw new ImmediateReturnError(msg, {error: msg});
}

function filenameFromUri(req) {
  var path = req.path;

  return path;
}

function makeAbsolute(filename, app) {
  var absfile = path.join(app.locals.config.dirs.datadir, filename);
  if (absfile.indexOf(app.locals.config.dirs.datadir) != 0) {
    errorExit('Illegal filename ' + filename);
  }
  absfile = path.resolve(absfile); // Resolve to absolute path

  return absfile;
}

function touchUserDir(userdir) {
  try {
    fs.utimesSync(userdir, Date.now(), Date.now());
  }
  catch (e) { }
}

function filenameuser(filename) {
  var m = filename.match(/^([\w][\w\.\-]*)(?:\/.*)?$/);

  return (m) ? m[1] : null;
}

function removeDirsSync(dirStart) {
  for (var dir = dirStart; ; dir = path.dirname(dir)) {
    if (fs.readdirSync(dir).length > 0) {
      // Directory not empty, we're done.
      return;
    }

    fsExtra.remove(dir);
  }
}

function isValidKey(user, key, app) {
  //
  // keydir is the directory containing the hashed user password.
  // It's a subdir off the user home directory called '.key'.  
  // Contents of this directory are files that are named
  // with the hashed user password
  // 

  var keydir = getKeyDir(user, app);
  var statObj = null;

  try {
    statObj = fs.statSync(keydir);
  }
  catch (e) {
  }

  if (!statObj) {
    //
    // .key directory does not seem to exist off user home directory.  
    // Assume that this means user does not have a password set.  
    //
    return true;
  }

  if (!statObj.isDirectory()) { 
    return true;
  }

  // Now we know its a directory
  var keys = fs.readdirSync(keydir);
  if (!keys || keys.length == 0) {
    // No key files, must mean no password for user.
    // So assume that this is valid.
    return true;
  }

  if (key) {
    for (var i = 0; i < keys.length; i++) {
      //
      // Password files are named with 'k' + the hashed password.  
      // See doSetKey() for implementation.  So check the substring
      // offset with 1 to ignore the starting 'k'
      //
      if (key.indexOf(keys[i].substring(1)) == 0) {
        return true;
      }
    }
  }
}

//
// Create a file with the hashed user password in the 
// user key dir.  This is called when the user sets or changes
// their password. 
//
function doSetKey(user, oldkey, newkey, res, app) {
  if (oldkey == newkey) {
    return;
  }

  var keydir = getKeyDir(user, app);

  try {
    // Create directory if not present
    if (!fs.existsSync(keydir)) {
      checkReservedUser(user, app);
      fsExtra.mkdirsSync(keydir);
    }

    if (oldkey) {
      //
      // Delete old password file if present
      //

      var keys = fs.readdirSync(keydir);

      for (var i = 0; i < keys.length; i++) {
        if (oldkey.indexOf(keys[i].substring(1)) == 0) {
          fs.unlink(path.join(keydir, keys[i]));
        }
      }
    }

    if (newkey) {
      // Now create new password file
      keyfile = path.join(keydir, 'k' + newkey);
      fs.closeSync(fs.openSync(keyfile, 'w'));
    }
  }
  catch (e) {
    errorExit('Could not set key.');
  }
}

function ImmediateReturnError(message, jsonObj) {
  this.name = 'ImmediateReturnError';
  this.message = message;
  this.jsonObj = jsonObj;
}
ImmediateReturnError.prototype = Error.prototype;

//
// Return user key directory where password is stored.
//
function getKeyDir(user, app) {
  return path.resolve(path.join(app.locals.config.dirs.datadir, user, '.key'));
}

//
// Return user home dir
//

function getUserHomeDir(user, app) {
  return path.resolve(path.join(app.locals.config.dirs.datadir, user));
}

function checkReservedUser(user, app) {
  var datadirAbs = path.resolve(app.locals.config.dirs.datadir);

  if (fs.existsSync(path.join(datadirAbs, user))) {
    return;
  }

  if (user != user.toLowerCase()) {
    errorExit('Username should be lowercase.');
  }
  
  var normalized = user.toLowerCase();
  if (fs.existsSync(path.join(datadirAbs, normalized))) {
    errorExit('Username is reserved.');
  }

  // Also check possible variations of badwords
  var normalizedi = translate(normalized, '013456789', 'oieasbtbg');
  if (normalized != normalizedi && 
      fs.existsSync(path.join(datadirAbs, normalizedi))) {
    errorExit('Username is reserved.');
  }

  var normalizedl = translate(normalized, '013456789', 'oleasbtbg');
  if (normalizedl != normalized &&
      fs.existsSync(path.join(datadirAbs, normalizedl))) {
    errorExit('Username is reserved.');
  }

  var checkwords = [normalized, normalizedi, normalizedl];
  var badwords = 
      fs.readFileSync(path.join(__dirname, 'bad-words.txt'), 'utf8').split(/\n/);
  var badsubstrings = 
      fs.readFileSync(path.join(__dirname, 'bad-substrings.txt'), 'utf8').split(/\n/); 

  for (var i = 0; i < checkwords.length; i++) {
    for (var j = 0; j < badwords.length; j++) {
      if (badwords[j].length > 0 && checkwords[i] == badwords[j]) {
        errorExit('Username is reserved.');
      }
    }
    for (var j = 0; j < badsubstrings.length; j++) {
      if (badsubstrings[j].length > 0 && 
	  checkwords[i].indexOf(badsubstrings[j]) != -1) {
        errorExit('Username is reserved.');
      }
    }
  }
}

function translate(source, from, to) {
  var fromArr = from.split('');
  var toArr = to.split('');
  var copy = new String(source);

  if (fromArr.length != toArr.length) {
    errorExit('Uh oh, parameters to translate are incorrect.');
  }

  for (var i = 0; i < fromArr.length; i++) {
    var x = copy.indexOf(fromArr[i]);
    if (x != -1) {
      // Match found, so replace it
      copy[x] = toArr[i];
    }
  }
  return copy;
}
